File: tests/tamm/Test_DLPNO_Ops.cpp
#include "macdecls.h"
#include "mpi.h"
#include <chrono>
#include <tamm/tamm.hpp>

using namespace tamm;

using DependencyMap = std::map<IndexVector, TiledIndexSpace>;

DependencyMap construct_full_dependency_2(const TiledIndexSpace &in_tis1,
                                          const TiledIndexSpace &in_tis2,
                                          const TiledIndexSpace &out_tis) {
  DependencyMap result;
  for (const auto &tile_1 : in_tis1) {
    for (const auto &tile_2 : in_tis2) {
      result.insert({{tile_1, tile_2}, out_tis});
    }
  }
  return result;
}

template <typename T>
void dlpno_T1_T2_allocate(Scheduler &sch, size_t N, Tile tilesize) {

  TiledIndexSpace LMO{IndexSpace{range(N)}, tilesize};
  DependencyMap depMO = construct_full_dependency_2(LMO, LMO, LMO);
  TiledIndexSpace PNO{LMO, {LMO, LMO}, depMO};

  auto [i, j] = LMO.labels<2>("all");
  auto [a, e] = PNO.labels<2>("all");

  Tensor<T> t1{i, e(i, i)};
  Tensor<T> t2{i, j, a(i, j), e(i, j)};

  sch.allocate(t1, t2).execute();
}

int main(int argc, char *argv[]) {

  tamm::initialize(argc, argv);

  int mpi_rank;
  MPI_Comm_rank(GA_MPI_Comm(), &mpi_rank);
#ifdef USE_TALSH
  TALSH talsh_instance;
  talsh_instance.initialize(mpi_rank);
#endif

  ProcGroup pg = ProcGroup::create_coll(GA_MPI_Comm());
  MemoryManagerGA *mgr = MemoryManagerGA::create_coll(pg);
  Distribution_NW distribution;
  RuntimeEngine re;
  ExecutionContext ec{pg, &distribution, mgr, &re};

  Scheduler sch{ec};

  dlpno_T1_T2_allocate<double>(sch, 10, 2);

#ifdef USE_TALSH
  talsh_instance.shutdown();
#endif

  tamm::finalize();

  return 0;
}

File: tests/python/Test_DLPNO_Ops.py
import pytamm as tamm


def construct_full_dependency_2(in_tis1, in_tis2, out_tis):
    result = {}

    for tile_1 in in_tis1:
        for tile_2 in in_tis2:
            result[(int(tile_1), int(tile_2))] = out_tis

    return result


def dlpno_T1_T2_allocate(sch, N, tilesize):
    LMO = tamm.TiledIndexSpace(
        tamm.IndexSpace(tamm.range(N)),
        tilesize,
    )

    depMO = construct_full_dependency_2(LMO, LMO, LMO)

    PNO = tamm.TiledIndexSpace(
        LMO,
        [LMO, LMO],
        depMO,
    )

    i, j = LMO.labels("all", count=2)
    a, e = PNO.labels("all", count=2)

    t1 = tamm.TensorDouble([i, e(i, i)])
    t2 = tamm.TensorDouble([i, j, a(i, j), e(i, j)])

    sch.allocate(t1, t2).execute()

    return t1, t2


def main():
    tamm.initialize(["dlpno_T1_T2_allocate.py"], False)

    try:
        pg = tamm.ProcGroup.create_world_coll()
        ec = tamm.ExecutionContext(
            pg,
            tamm.DistributionKind.nw,
            tamm.MemoryManagerKind.ga,
        )

        sch = tamm.Scheduler(ec)

        dlpno_T1_T2_allocate(sch, 10, 2)

    finally:
        tamm.finalize(True)


if __name__ == "__main__":
    main()